paper/src/main/java/sh/okx/railswitch/switches/SwitchListener.java

```diff
@@ -84,46 +117,26 @@ public void onSwitchTrigger(BlockRedstoneEvent event) {
         }
         // If Citadel is enabled, check that the sign and the rail are on the same group
         if (CITADEL_GLUE.isSafeToUse()) {
-            if (!CITADEL_GLUE.doSignAndRailHaveSameReinforcement(above, block)) {
+            if (!CITADEL_GLUE.doSignAndRailHaveSameReinforcement(sign, rail)) {
                 return;
             }
         }
-        // Determine whether a player has a destination that matches one of the destinations
-        // listed on the switch signs, or match if there's a wildcard.
-        boolean matched = false;
-        String setDest = SettingsManager.getDestination(player);
-        if (!Strings.isNullOrEmpty(setDest)) {
-            String[] playerDestinations = setDest.split(" ");
-            String[] switchDestinations = Arrays.copyOfRange(lines, 1, lines.length);
-            matcher:
-            for (String playerDestination : playerDestinations) {
-                if (Strings.isNullOrEmpty(playerDestination)) {
-                    continue;
-                }
-                if (playerDestination.equals(WILDCARD)) {
-                    matched = true;
-                    break;
-                }
-                for (String switchDestination : switchDestinations) {
-                    if (Strings.isNullOrEmpty(switchDestination)) {
-                        continue;
-                    }
-                    if (switchDestination.equals(WILDCARD)
-                            || playerDestination.equalsIgnoreCase(switchDestination)) {
-                        matched = true;
-                        break matcher;
-                    }
-                }
-            }
-        }
-        switch (type) {
-            case NORMAL:
-                event.setNewCurrent(matched ? 15 : 0);
-                break;
-            case INVERTED:
-                event.setNewCurrent(matched ? 0 : 15);
-                break;
+        
+        //Do the rail switching
+        String dest_string = SettingsManager.getDestination(player);
+        try {
+            if (logic.decide(player)) event.setNewCurrent(15);
+            else event.setNewCurrent(0);
+        } catch (Exception e) {
+            emit_failure_smoke(sign.getLocation());
+            event.setNewCurrent(0);
         }
     }
 
+    /** Emit a campfire particle at the location of the sign to indicate failure */
+    public void emit_failure_smoke(Location block_coordinates) {
+        Location location = block_coordinates.add(0.5, 0.5, 0.5);
+        location.getWorld().spawnParticle(Particle.CAMPFIRE_COSY_SMOKE, location, 0, 0.0, 0.0, 0.0, 3.0);
+    }
```

Comment:
Setting aside the snake case for a moment, you can readily one-liner this with:
```java
public void emit_failure_smoke(Location block_coordinates) {
    block_coordinates.getWorld().spawnParticle(block_coordinates.toCenterLocation(), 0, 0.0, 0.0, 0.0, 3.0);
}
```
[API Reference](https://jd.papermc.io/paper/1.18/org/bukkit/Location.html#toCenterLocation())